### ISSUE TITLE
Performance of broadcasting scalars over views is order-dependent

### ISSUE THREAD

**Issue by GravityAssisted** [JuliaLang/julia#22165]
Consider these two simple examples. ```work``` is a vector by ```byr1``` is a scalar. I am trying to achieve a non-allocating array assignment.

```
work = rand(6)
byr1 = 0.4
julia> @benchmark @inbounds @views $work[4:6]  .= $byr1.*$work[4:6]
BenchmarkTools.Trial: 
  memory estimate:  48 bytes
  allocs estimate:  1
  --------------
  minimum time:     18.073 ns (0.00% GC)
  median time:      18.703 ns (0.00% GC)
  mean time:        24.711 ns (20.62% GC)
  maximum time:     2.557 μs (97.40% GC)
  --------------
  samples:          10000
  evals/sample:     997

julia> @benchmark @inbounds @views $work[4:6]  .= $work[4:6].*$byr1
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     8.319 ns (0.00% GC)
  median time:      8.599 ns (0.00% GC)
  mean time:        8.725 ns (0.00% GC)
  maximum time:     25.867 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
```

Why is the first case allocating while the second case is not ? Looks like the allocation happens only if I multiple byr1 before work array. These seems like a bug or maybe I missed something here?

thanks

**Comment by GravityAssisted** [comment]
More investigation. If I replace the right hand side ```work``` array view with an array ```r``` then this problem goes away.
```
r = rand(3)

julia> @benchmark @inbounds @views $work[4:6]  .= $r.*$byr1
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     9.858 ns (0.00% GC)
  median time:      11.395 ns (0.00% GC)
  mean time:        10.942 ns (0.00% GC)
  maximum time:     30.621 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999

julia> @benchmark @inbounds @views $work[4:6]  .= $byr1.*$r
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     9.717 ns (0.00% GC)
  median time:      10.137 ns (0.00% GC)
  mean time:        10.127 ns (0.00% GC)
  maximum time:     27.685 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
```
So it seems there is something going on with the associativity of the view with a scalar ?

**Comment by KristofferC** [comment]
Both examples does 2 allocations now (1 in the case of providing `r`).

**Comment by laborg** [comment]
This seems to be fixed (Julia 1.9.3):
```
julia> @btime @inbounds @views $work[4:6]  .= $byr1.*$work[4:6]
  11.196 ns (0 allocations: 0 bytes)
3-element view(::Vector{Float64}, 4:6) with eltype Float64:
 0.0
 0.0
 0.0

julia> @btime  @inbounds @views $work[4:6]  .= $work[4:6].*$byr1
  11.200 ns (0 allocations: 0 bytes)
3-element view(::Vector{Float64}, 4:6) with eltype Float64:
 0.0
 0.0
 0.0
```
